packages/flutter/lib/src/rendering/flex.dart

```diff
@@ -505,93 +591,47 @@ class RenderFlex extends RenderBox with ContainerRenderObjectMixin<RenderBox, Fl
+      Axis.vertical   => BoxConstraints(maxWidth: extent),
+    };
     if (_direction == sizingDirection) {
       // INTRINSIC MAIN SIZE
-      // Intrinsic main size is the smallest size the flex container can take
-      // while maintaining the min/max-content contributions of its flex items.
-      double totalFlex = 0.0;
-      double inflexibleSpace = 0.0;
-      double maxFlexFractionSoFar = 0.0;
-      RenderBox? child = firstChild;
-      while (child != null) {
-        final int flex = _getFlex(child);
-        totalFlex += flex;
-        if (flex > 0) {
-          final double flexFraction = childSize(child, extent) / _getFlex(child);
-          maxFlexFractionSoFar = math.max(maxFlexFractionSoFar, flexFraction);
-        } else {
-          inflexibleSpace += childSize(child, extent);
+      Size computeIntrinsicMainExtentForChild(RenderBox child, BoxConstraints constraints) {
+        switch (_direction) {
+          case Axis.horizontal:
+            final double height = constraints.maxHeight;
+            return Size(childSize(child, height), height);
+          case Axis.vertical:
+            final double width = constraints.maxWidth;
+            return Size(width, childSize(child, width));
         }
-        final FlexParentData childParentData = child.parentData! as FlexParentData;
-        child = childParentData.nextSibling;
       }
-      return maxFlexFractionSoFar * totalFlex + inflexibleSpace;
+      return _computeSizes(constraints: extentConstraints, layoutChild: computeIntrinsicMainExtentForChild, computeBaseline: ChildLayoutHelper.getDryBaseline).size.mainAxisExtent;
     } else {
       // INTRINSIC CROSS SIZE
-      // Intrinsic cross size is the max of the intrinsic cross sizes of the
-      // children, after the flexible children are fit into the available space,
-      // with the children sized using their max intrinsic dimensions.
-
-      // Get inflexible space using the max intrinsic dimensions of fixed children in the main direction.
-      final double availableMainSpace = extent;
-      int totalFlex = 0;
-      double inflexibleSpace = 0.0;
-      double maxCrossSize = 0.0;
-      RenderBox? child = firstChild;
-      while (child != null) {
-        final int flex = _getFlex(child);
-        totalFlex += flex;
-        late final double mainSize;
-        late final double crossSize;
-        if (flex == 0) {
-          switch (_direction) {
-            case Axis.horizontal:
-              mainSize = child.getMaxIntrinsicWidth(double.infinity);
-              crossSize = childSize(child, mainSize);
-            case Axis.vertical:
-              mainSize = child.getMaxIntrinsicHeight(double.infinity);
-              crossSize = childSize(child, mainSize);
-          }
-          inflexibleSpace += mainSize;
-          maxCrossSize = math.max(maxCrossSize, crossSize);
+      Size computeIntrinsicCrossExtentForChild(RenderBox child, BoxConstraints constraints) {
```

Comment:
nit: these names seem inaccurate. It does not only return the extend, it returns a full size.

FWIW, I found this code easier to follow when it was implemented in terms of mainExtend and crossExtend and not in terms of width/height.